### PR TITLE
[GPOS + Market] Add mutex to balance

### LINKED ISSUE #164: [GPoS+Market] Pledge and bond should be mutex

### PR DESCRIPTION
**Pull request by MyronFanQiu**
Fix #164

### CODE REVIEW

File: cstrml/staking/src/lib.rs [hunk 1]
```diff
@@ -1946,6 +1948,24 @@ impl<T: Trait> Module<T> {
...
+pub trait EnhancedBalanceInterface<AccountId, Balance>: system::Trait {
+    fn free_balance_for_staking_only(who: &AccountId) -> Balance;
+}
+
+impl<T: Trait> EnhancedBalanceInterface<<T as system::Trait>::AccountId, BalanceOf<T>> for T where T: balances::Trait
+{
+    fn free_balance_for_staking_only(who: &<T as system::Trait>::AccountId) -> BalanceOf<T>
+    {
+        let to_balanceof = |e: T::Balance| <T::CurrencyToVote as Convert<u128, BalanceOf<T>>>::convert(TryInto::<u128>::try_into(e).ok().unwrap());
+        let sum = <balances::Module<T>>::locks(who).iter().fold(Zero::zero(), |acc: <T as balances::Trait>::Balance, value| {
```
Comment (MyronFanQiu) [inline cstrml/staking/src/lib.rs:1960]: There are two concerns for such design. 
- `locks` is not in any trait and cannot be used directly. I have to include balance as one dependency and extend its ability.
- There are two types `T::Balance` and `BalanceOf<T>`. I have to use u128 to covert one to another.

Comment (MyronFanQiu) [reply cstrml/staking/src/lib.rs:1960]: Another option is to include `Balance` module and modify it directly.

File: cstrml/staking/src/lib.rs [hunk 2]
```diff
@@ -359,6 +359,8 @@ pub trait Trait: frame_system::Trait {
     /// The staking balance.
     type Currency: LockableCurrency<Self::AccountId, Moment = Self::BlockNumber>;
 
+    type EnhancedBalanceInterface: EnhancedBalanceInterface<Self::AccountId, BalanceOf<Self>>;
```
Comment (zikunfan) [inline cstrml/staking/src/lib.rs:362]: Enhanced seems unclear to me

File: cstrml/staking/src/lib.rs (hunk 1 above)
Comment (zikunfan) [reply cstrml/staking/src/lib.rs:1960]: As for `Maxwell 1.0`, we take ***solution1***, import balance should be a long-term solution. And `locks` should be and only be used in `staking` and `market` pallets inside crust repo.

Comment (MyronFanQiu) [reply cstrml/staking/src/lib.rs:1960]: Discussed offline, choose `solution2` since we need balance in the future anyway.

Comment (zikunfan) [reply cstrml/staking/src/lib.rs:1960]: > As for `Maxwell 1.0`, we take ***solution1***, import balance should be a long-term solution. And `locks` should be and only be used in `staking` and `market` pallets inside crust repo.

Talked offline, considering the long term repeated work, we take sol2 as our final offer.

**Pull request closed without merging**